// app/utils/tests.js
import * as _ from 'lamb';

export const testResultsURL = 'https://gist.githubusercontent.com/NestaTestUser/01b59a18ffe2a87e300b88f0be973ad0/raw/Browserstack%20test%20results%20for%20BEIS%20Indicators%20webapp';

const osMap = {
	'Windows': 'Windows',
	'macOS': 'OS X'
};
const browserMap = {
	'Microsoft Edge': 'Edge',
	'Chrome': 'Chrome',
	'Safari': 'Safari',
	'Firefox': 'Firefox'
};

const getOS = _.getPath('capabilities.bstack:options.os');
const getVersion = _.getPath('capabilities.bstack:options.osVersion');
const getBrowser = _.getPath('capabilities.browserName');
const getBrowserV = _.getPath('capabilities.browserVersion');

export const groupTests = _.pipe([
	_.groupBy(getOS),
	_.mapValuesWith(_.pipe([
		_.groupBy(getVersion),
		_.mapValuesWith(_.pipe([
			_.groupBy(getBrowser),
			_.mapValuesWith(_.groupBy(getBrowserV)),
		])),
	])),
]);

export const getTest = (testsIndex, env) => {
	const os = osMap[env.os.name];
	const osVersion = env.os.versionName;
	const browser = browserMap[env.browser.name];
	const browserVersion = env.browser.version.split('.').slice(0,2).join('.');
	const results = testsIndex?.[os]?.[osVersion]?.[browser]?.[browserVersion];
	return results?.[0];
};

export const summarizeResults = (platform) => {
	if (platform) {
		return {
			tested: true,
			passed: _.reduce(
				platform.results,
				(previousPassed, test) => previousPassed && test.result.passed[1]
			),
			// TODO add reasons for failure
			reasonsForFailure: [],
		};
	} else {
		return {
			tested: false,
			passed: false,
			reasonsForFailure: [`Untested platform.`],
		};
	}
};
