// app/config.js
import regionSpecs from '@svizzle/atlas/src/specs';

export const toolName = 'BEIS/Nesta R&D dashboard';
export const toolLongName = 'BEIS/Nesta Research & Development spatial data tool';
export const contactEmail = 'dataanalytics@nesta.org.uk';

/* navigation */

export const hrefBase = '/indicators';
export const userTestingUrl = 'https://docs.google.com/forms/d/e/1FAIpQLSekqhPhm4VNDdQNmpbE0EjTQj8zruLw8mIyqJH5hV9z-Dytbw/viewform?embedded=true';

/* flags */

export const flags = {
	showPOIs: true
}

/* regional selection */

const regionType = 'NUTS';

export const regionSettings = {
	canSelectLevels: false,
	filterableLevel: 1,
	ignoredRegions: [],
	key: 'NUTS_ID',
	level: 2,
	level0: 'UK',
	levels: regionSpecs[regionType].levels,
	objectId: 'NUTS',
	resolution: '03M',
	type: regionType,
}

/* testing */

export const urlBases = {
    development: 'http://localhost:3000',
    preview: 'https://beis-indicators-dev.netlify.app',
    // production: '',
};

export const lighthouseUrls = {
	Home: '/',
	Methodology: '/methodology',
	Guide: '/guide',
	Indicators: '/indicators',
	Trends: '/indicators/area_university_site',
	Geo: '/indicators/area_university_site/2015',
};

/*
There's a bug in Lighthouse 7.4.0 which causes the audit on the following
pages to fail when running in mobile display emulation. Commenting them
for now as tests are being run in desktop mode only pending the fix.
Ref.: https://github.com/GoogleChrome/lighthouse/issues/12039
*/
export const failingA11yAudit = [];

export const isDev = process.env.NODE_ENV === 'development';
