// app/stores.js
import { derived, writable } from 'svelte/store';
import { linearScale } from 'yootils';
import {transformValues, mergeObj, mergeWithMerge} from '@svizzle/utils';
import * as _ from 'lamb';

import {lookup, yearExtent} from './data/groups';
import UK_NUTS_1_2 from './data/UK_NUTS_1_2.json';

const toggle = x => !x;

/* groups */

export const lookupStore = writable(lookup);

/* layout */

export const safetyDefault = {left: 80, right: 80};
export const safetyStore = writable(safetyDefault);
export const resetSafetyStore = () => {
	safetyStore.set(safetyDefault);
}

export const timelineHeightStore = writable(0);
export const timelineWidthStore = writable(0);

export const timelineLayoutStore = derived([
	timelineHeightStore,
	timelineWidthStore,
	safetyStore
], ([height, width, safety]) => {
	const padding = 10;
	const fontSize = Math.min(height / 4, 14);
	const radius = Math.min(height / 8, 7);
	const fullExtent = [
		padding + radius,
		width - padding - radius
	];
	const fullScaleX = linearScale(yearExtent, fullExtent);
	const start = fullExtent[0] + safety.left;
	const end = fullExtent[1] - safety.right;
	const scaleX = linearScale(yearExtent, [start, end]);
	const y1 = height / 3;
	const y2 = (height + y1 + radius) / 2;
	const ym = height / 2;

	return {
		end,
		fontSize,
		fullScaleX,
		height,
		radius,
		scaleX,
		start,
		width,
		y1,
		y2,
		ym,
	}
});

/* selection */

export const selectedYearStore = writable();
export const resetSelectedYear = () => {
	selectedYearStore.set();
}

export const availableYearsStore = writable([]);

export const resetSelection = () => {
	availableYearsStore.set([]);
	resetSelectedYear();
};

/* regional selection */

// selected regions

const allNutsSelected =
	_.mapValues(UK_NUTS_1_2, mergeObj({selected: true}));
const noNutsSelected =
	_.mapValues(UK_NUTS_1_2, mergeObj({selected: false}));
export const nutsSelectionStore = writable(allNutsSelected);
export const selectAllRegions = () => {
	nutsSelectionStore.set(allNutsSelected);
};
export const deselectAllRegions = () => {
	nutsSelectionStore.set(noNutsSelected);
};
export const toggleRegionNUTS1 = id => {
	nutsSelectionStore.update(_.updatePath(`${id}.selected`, toggle));
};
export const selectRegionNUTS1 = id => {
	nutsSelectionStore.update(_.setPath(`${id}.selected`, true));
};

const getSelectedChildren = _.pipe([
	_.values,
	_.filterWith(_.getKey('selected')),
	_.flatMapWith(_.getKey('children'))
]);
export const selectedNUT2IdsStore = derived(
	[nutsSelectionStore],
	([nutsSelection]) => getSelectedChildren(nutsSelection)
);

// preselected regions

const getNUTS1Children = id => UK_NUTS_1_2[id].children;
export const preselectedNUTS1IdStore = writable(null);
export const preselectedNUTS2IdsStore = derived(
	[preselectedNUTS1IdStore],
	([id]) => id && getNUTS1Children(id) || []
);

// modal utils

const hidden = {isVisible: false};
const toggleIsVisible = transformValues({isVisible: toggle});

// modal

export const geoModalStore = writable(hidden);
export const hideGeoModal = () => {
	geoModalStore.set(hidden);
};
export const toggleGeoModal = () => {
	geoModalStore.update(toggleIsVisible);
};

// selection mode

export const doFilterRegionsStore = writable(false);

/* info modal */

export const infoModalStore = writable(hidden);
export const hideInfoModal = () => {
	infoModalStore.set(hidden);
};
export const toggleInfoModal = () => {
	console.log('.....')
	infoModalStore.update(toggleIsVisible);
};
