// app/stores.js
import { derived, writable } from 'svelte/store';
import { linearScale } from 'yootils';

import {lookup, yearExtent} from './data/groups';

/* groups */

export const lookupStore = writable(lookup);

/* layout */

const safety = 10;

export const timelineHeightStore = writable(0);
export const timelineWidthStore = writable(0);
export const timelineLayoutStore = derived([
  timelineHeightStore,
  timelineWidthStore
], ([height, width]) => {
  const fontSize = Math.min(height / 4, 14);
  const radius = Math.min(height / 8, 7);
  const start = radius + safety;
  const end = width - start;
  const scaleX = linearScale(yearExtent, [start, end]);
  const y1 = height / 3;
  const y2 = (height + y1 + radius) / 2;
  const ym = height / 2;

  return {height, width, fontSize, radius, start, end, scaleX, y1, y2, ym}
});

/* selection */

export const selectedYearStore = writable();

export const availableYearsStore = writable([]);
export const resetSelection = () => {
  availableYearsStore.set([]);
  selectedYearStore.set();
};
